src/pyodide/internal/asgi.py

```diff
@@ -99,53 +99,118 @@ async def send(got):
             headers = [(k.decode(), v.decode()) for k, v in got["headers"]]
-        if got["type"] == "http.response.body":
-            # intentionally leak body to avoid a copy
-            #
-            # Apparently the `Response` constructor will not eagerly copy a
-            # TypedArray argument, so if we don't leak the argument the response
-            # body is corrupted. This is fine because the session is going to
-            # exit at this point.
-            px = create_proxy(got["body"])
+            # Check if this is a server-sent events response
+            for k, v in headers:
+                if k.lower() == "content-type" and v.lower().startswith(
+                    "text/event-stream"
+                ):
+                    is_sse = True
+
+                    # For SSE, create and return the response immediately after http.response.start
+                    resp = Response.new(
+                        readable, headers=Object.fromEntries(headers), status=status
+                    )
+                    result.set_result(resp)
+                    break
+
+        elif got["type"] == "http.response.body":
+            body = got["body"]
+            more_body = got.get("more_body", False)
+
+            # Convert body to JS buffer
+            px = create_proxy(body)
             buf = px.getBuffer()
             px.destroy()
-            resp = Response.new(
-                buf.data, headers=Object.fromEntries(headers), status=status
-            )
-            result.set_result(resp)
 
-    await app(request_to_scope(req, env), receive, send)
-    return await result
+            if is_sse:
+                # For SSE, write chunk to the stream
+                await writer.write(buf.data)
+                # If this is the last chunk, close the writer
+                if not more_body:
+                    await writer.close()
+                    finished_response.set()
+            else:
+                resp = Response.new(
+                    buf.data, headers=Object.fromEntries(headers), status=status
+                )
+                result.set_result(resp)
+                finished_response.set()
+
+    # Run the application in the background to handle SSE
+    async def run_app():
+        try:
+            await app(request_to_scope(req, env), receive, send)
+
+            # If we get here and no response has been set yet, the app didn't generate a response
+            if not result.done():
+                await writer.close()  # Close the writer
+                finished_response.set()
+                result.set_exception(
+                    RuntimeError("The application did not generate a response")
+                )
+        except Exception as e:
+            # Handle any errors in the application
+            if not result.done():
+                await writer.close()  # Close the writer
+                finished_response.set()
+                result.set_exception(e)
```

Comment:
A minor simplification:
```suggestion
            if not result.done():
                raise RuntimeError("The application did not generate a response")
        except Exception as e:
            # Handle any errors in the application
            if not result.done():
                await writer.close()  # Close the writer
                finished_response.set()
                result.set_exception(e)
```